util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,428 @@
+        secded_width = self.lc_state.config['secded']['data_width']
+        if otp_width != secded_width:
+            raise RuntimeError('OTP width and SECDED data width must be equal')
+
+        if 'seed' not in img_config:
+            raise RuntimeError('Missing seed in configuration.')
+
+        img_config['seed'] = check_int(img_config['seed'])
+        log.info('Seed: {0:x}'.format(img_config['seed']))
+        log.info('')
+
+        # Re-initialize with seed to make results reproducible.
+        random.seed(OTP_IMG_SEED_DIVERSIFIER + img_config['seed'])
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Adding values to {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+        # Key accounting
+        img_config_check = img_config.copy()
+        del img_config_check['seed']
+        del img_config_check['partitions']
+        _check_unused_keys(img_config_check, 'in image config')
+
+        log.info('')
+        log.info('Parsing OTP image successfully completed.')
+
+    def merge_part_data(self, part):
+        '''This validates and merges the partition data into the memory map dict'''
+
+        part.setdefault('items', [])
+        if not isinstance(part['items'], list):
+            raise RuntimeError('the "items" key must contain a list')
+
+        # Check if partition name exists in memory map
+        part.setdefault('name', 'unknown_name')
+        mmap_part = self.get_part(part['name'])
+        if mmap_part is None:
+            raise RuntimeError('Partition {} does not exist'.format(
+                part['name']))
+
+        # Only partitions with a hardware digest can be locked.
+        part.setdefault('lock', 'false')
+        part['lock'] = check_bool(part['lock'])
+        if part['lock'] and not \
+           mmap_part['hw_digest']:
+            raise RuntimeError(
+                'Partition {} does not contain a hardware digest'.format(
+                    part['name']))
+
+        # Augment memory map datastructure with lock bit.
+        mmap_part['lock'] = part['lock']
+
+        if part['name'] == 'LIFE_CYCLE':
+            part.setdefault('state', 'RAW')
+            part.setdefault('count', 0)
+
+            part['count'] = check_int(part['count'])
+            if len(part['items']) > 0:
+                raise RuntimeError(
+                    'Life cycle items cannot directly be overridden')
+            if part['lock']:
+                raise RuntimeError('Life cycle partition cannot be locked')
+
+            # Augment life cycle partition with correct life cycle encoding
+            state = self.lc_state.encode('lc_state', str(part['state']))
+            count = self.lc_state.encode('lc_cnt', str(part['count']))
```

Comment:
Do we want to add a check here that `lc_cnt` 0 only applies to `RAW` state?

Reply:
good idea, let me add that